charger/em2go-home.go

```diff
@@ -0,0 +1,335 @@
+	"encoding/binary"
+	"fmt"
+	"time"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/modbus"
+	"github.com/volkszaehler/mbmd/meters/rs485"
+)
+
+// https://www.em2go.de/download2/ModBus TCP Registers EM2GO Home Series.pdf
+
+// Em2Go charger implementation
+type Em2GoHome struct {
+	log  *util.Logger
+	conn *modbus.Connection
+}
+
+const (
+	em2GoHomeRegStatus			= 0   // Uint16 RO ENUM
+	em2GoHomeRegConnectorState	= 2   // Uint16 RO ENUM
+	em2GoHomeRegErrorCode		= 4   // Uint16 RO ENUM
+	em2GoHomeRegCurrent1		= 6   // Uint16 RO 0.1A
+	em2GoHomeRegCurrent2		= 8   // Uint16 RO 0.1A
+	em2GoHomeRegCurrent3		= 10  // Uint16 RO 0.1A
+	em2GoHomeRegPower			= 12  // Uint32 RO 1W
+	em2GoHomeRegPower1			= 16  // Uint16 RO 1W
+	em2GoHomeRegPower2			= 18  // Uint16 RO 1W
+	em2GoHomeRegPower3			= 20  // Uint16 RO 1W
+	em2GoHomeRegEnergy			= 28  // Uint16 RO 0.1KWh
+	em2GoHomeRegMaxCurrent		= 32  // Uint16 RO 0.1A
+	em2GoHomeRegMinCurrent		= 34  // Uint16 RO 0.1A
+	em2GoHomeRegCableMaxCurrent	= 36  // Uint16 RO 0.1A
+	em2GoHomeRegSerial			= 38  // Chr[16] RO UTF16
+	em2GoHomeRegChargedEnergy	= 72  // Uint16 RO 0.1kWh
+	em2GoHomeRegChargeDuration	= 78  // Uint32 RO 1s
+	em2GoHomeRegSafeCurrent		= 87  // Uint16 WR 0.1A
+	em2GoHomeRegCommTimeout		= 89  // Uint16 WR 1s
+	em2GoHomeRegCurrentLimit	= 91  // Uint16 WR 0.1A
+	em2GoHomeRegChargeMode		= 93  // Uint16 WR ENUM
+	em2GoHomeRegChargeCommand	= 95  // Uint16 WR ENUM
+	em2GoHomeRegVoltage1		= 109 // Uint16 RO 0.1V
+	em2GoHomeRegVoltage2		= 111 // Uint16 RO 0.1V
+	em2GoHomeRegVoltage3		= 113 // Uint16 RO 0.1V
+	em2GoHomeRegPhases			= 200 // Set charging phase 1 unsigned
+)
+
+func init() {
+	registry.Add("em2go-home", NewEm2GoHomeFromConfig)
+}
+
+//go:generate go run ../cmd/tools/decorate.go -f decorateEm2GoHome -b *Em2GoHome -r api.Charger -t "api.PhaseSwitcher,Phases1p3p,func(int) error"
+
+// NewEm2GoHomeFromConfig creates a Em2Go charger from generic config
+func NewEm2GoHomeFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := modbus.TcpSettings{
+		ID: 255,
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewEm2GoHome(cc.URI, cc.ID)
+}
+
+// NewEm2GoHome creates Em2GoHome charger
+func NewEm2GoHome(uri string, slaveID uint8) (api.Charger, error) {
+	uri = util.DefaultPort(uri, 502)
+
+	conn, err := modbus.NewConnection(uri, "", "", 0, modbus.Tcp, slaveID)
+	if err != nil {
+		return nil, err
+	}
+	
+	// Add delay of 60 miliseconds between requests
+	conn.Delay(60 * time.Millisecond)
+	
+	log := util.NewLogger("em2go-home")
+	conn.Logger(log.TRACE)
+
+	wb := &Em2GoHome{
+		log:  log,
+		conn: conn,
+	}
+
+	_, v2, v3, err := wb.Voltages()
+
+	var phases1p3p func(int) error
+	if v2 != 0 && v3 != 0 {
+		wb.log.DEBUG.Println("detected 3p em2go-home")
+		phases1p3p = wb.phases1p3p
+	} else {
+		wb.log.DEBUG.Println("detected 1p em2go-home")
+	}
+
+	return decorateEm2GoHome(wb, phases1p3p), err
+}
```

Comment:
Braucht man das?
Wenn der Charger nur 1p kann wird ohnehin nicht umgeschaltet.

Reply:
Brauchen wir nicht, wenn das automatisch funktioniert. War nur angedacht, dies mitzuteilen, wenn der Charger mit einer Phase angebunden ist.

Reply:
Im Endeffekt stellt der Anwender das dann ohnehin in der UI ein und bei einphasig gibt es dann sowieso keine Umschaltung. Daher ist das eher redundant. Tut nicht weh aber bringt auch nix ;)